pkg/oc/cli/cmd/debug.go

```diff
@@ -241,10 +242,20 @@ func (o *DebugOptions) Complete(cmd *cobra.Command, f *clientcmd.Factory, args [
 		return fmt.Errorf("you must identify a resource with a pod template to debug")
 	}
 
-	template, err := f.ApproximatePodTemplateForObject(infos[0].Object)
-	if err != nil && template == nil {
-		return fmt.Errorf("cannot debug %s: %v", infos[0].Name, err)
+	var template *kapi.PodTemplateSpec
```

Comment:
Why wouldn't we always use ApproximatePodTemplateForObject for debug?

Reply:
ApproximatePodTemplateForObject shouldn't wait for pods under DC

Reply:
With a deployment object, the default case was being hit, as its type resolved to `*extensions.Deployment`. It then got stuck in the call to `f.AttachablePodForObject`. I could add a new case to the switch for `*extensions.Deployment` with the same logic as the `*deployapi.DeploymentConfig` case. wdyt?

Reply:
https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L209

Reply:
There should be an "approximatepodtemplateforobject" for RCs and RSs and Deployments and StatefulSets and Daemonsets upstream that doesn't depend on pods existing.  So I would expect to always be using approximate pod template and for it to never block.

Reply:
Without adding the case for `*extensions.Deployment`, current behavior hits the default case in https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L209. Then `AttachablePodForObject` method is [called on factory_object_mapping upstream with a timeout of 1 min](https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L210).

[Here is where the blocking seems to be happening](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory_object_mapping.go#L408).

Another possible solution could be to skip the call to `AttachablePodForObject` altogether if there are 0 replicas, and add a new type switch case for `*extensions.Deployment` [here](https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L217) that just returns `t.Spec.Template`

Reply:
Don't call AttachablePodForObject ever.  ApproximatePodTemplateforObject should always be used for debug all the way down, and nothing in APTFO should ever wait for a container to start (it doesn't have to).  attach requires a running pod, debug does not.